Add @automattic/privacy-toolset npm library

robertsreberski opened the pull request:
#### Proposed Changes

* Introduce a new reusable library for privacy components and tools
* Create a reusable Cookie Banner component optimized for SSR
<img width="1138" alt="CleanShot 2022-11-22 at 10 29 35@2x" src="https://user-images.githubusercontent.com/8419292/203368855-e4f46c67-c720-4527-82f5-7ee536f3e8c3.png">
<img width="1144" alt="CleanShot 2022-11-22 at 10 30 29@2x" src="https://user-images.githubusercontent.com/8419292/203368960-1d0247a2-6c2d-4f07-9d68-7e1f62aef622.png">


#### Testing Instructions

* Checkout the changes
* Run `yarn workspace @automattic/privacy-toolset run storybook`
* Storybook should automatically open in your browser
* Go to the path `?path=/story/cookie-banner--default` (should be open by default)
* The Cookie Banner should be styled as per the design above
* Click "Accept All" - it should result in the action `accept: {essential: true, analytics: true, advertising: true}`
* Click "Customize" - you should see an expanded version of the Cookie Banner as per the design above
* Uncheck the "Analytics" bucket
* Click "Accept Selection" - it should result in the action `accept: {essential: true, analytics: false, advertising: true}`

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #69304

robertsreberski committed:
Initialize @automattic/privacy-toolset repo

robertsreberski committed:
Add example TestComponent

robertsreberski committed:
Remove TestComponent

robertsreberski committed:
Add CookieBanner component

robertsreberski committed:
Update styles.scss to use tabs

robertsreberski committed:
Add storybook to CookieBanner

robertsreberski committed:
Add dev dependencies

robertsreberski committed:
Revert to currently used version of storybook

robertsreberski committed:
Run yarn dedupe

robertsreberski committed:
Merge branch 'trunk' into add/privacy-components-lib

robertsreberski committed:
Update readme and changelog

CodeyGuyDylan commented on packages/privacy-toolset/README.md:
```diff
@@ -0,0 +1,37 @@
+# Components
+
+A library of tools React components designed to unify privacy mechanisms in Automattic products.
```
This could be reworded, I think sounds a bit weird like this. Maybe:
`A library of tools created as React components designed to unify privacy mechanism in Automattic products.`

CodeyGuyDylan commented on packages/privacy-toolset/README.md:
```diff
@@ -0,0 +1,37 @@
+
+### Tests
+
+N/A
```
Are there plans to add tests for this in the future? I am a strong advocate of adding tests to any new code, especially something brand new like this 😄

robertsreberski replied:
I've left this section when we'll start adding more logic to here (this lib is created to have both react component and helper functions). I'm not sure if we should test such presentational react components? Or do you see a good testing case here?

CodeyGuyDylan replied:
On components such as these, I've seen test cases written to make sure the React state is handled correctly when certain actions are taken. Could be useful here! Not going to fight to say it's pertinent though

jboland88 replied:
I would echo that some tests would be useful since this is a new package that will have some wider application. I think it could be useful to test:
- Output of simpleConsent shows as expected on initial render
- customConsent shows when clicking the "Customize" control
- The onAccept handler prop receives expected/  properly typed input from the component for both the "Accept All" action and a customized selection of preferences.

robertsreberski replied:
Thanks! I've implemented tests based on snapshots and listening to onAccept calls.

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/cookie-banner.stories.tsx:
```diff
@@ -0,0 +1,92 @@
+					or conducting analytics, <strong>marketing activities</strong> to help deliver relevant
+					marketing or content, and your <strong>user experience</strong> such as by remembering
+					your account name, language settings, or cart information, where applicable. You can
+					customize your cookie settings below. Learn more in our{ ' ' }
```
Just a comment on wording here. Twice there is the phrase `such as by` and I think just the word `by` would actually be better. For example:
`our website such as by improving security or conducting analytics`
would be
`our website by improving security or conducting analytics`

gmovr replied:
Just chiming in here. This wording is from legal, I'm sure we can change it down the line though, as it's minor and doesn't affect the messaging. :)

robertsreberski replied:
Thanks for the suggestion, I agree we can change it. However, the text provided here is just for presentational purposes; the real copy will be coming from the context of the implementation of this component.

I think I'll replace it with lorem ipsum here to not distract future contributors/reviewers from component code - which is the essence of this PR.

Btw, @gmovr, do you have some thoughts about having some universal copy stored somewhere in the library? Currently, we need to define it in all places we use a8c-analytics.js, and managing the updates might be troublesome in the future.

CodeyGuyDylan replied:
Got it!

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/cookie-banner.stories.tsx:
```diff
@@ -0,0 +1,92 @@
+import { action } from '@storybook/addon-actions';
+import { Story } from '@storybook/react';
+import React from 'react';
+import { CookieBanner, CookieBannerProps } from '.';
```
Here you can have
`import type { CookieBanner, CookieBannerProps } from '.';`
so Typescript knows to remove this when compiling

Same as line 2 where you import "Story"

CodeyGuyDylan replied:
I see this throughout all of the React TypeScript files, I won't litter the review with these comments, this is just something to improve in all of them 😄

robertsreberski replied:
Thanks, that's a good protip!

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/cookie-banner.stories.tsx:
```diff
@@ -0,0 +1,92 @@
+					<a href="https://automattic.com/privacy/" target="_blank" rel="noreferrer">
+						Privacy Policy
+					</a>{ ' ' }
+					and{ ' ' }
```
Personally I hate those `{ ' ' }` things, this is personal preference but I'd prefer
`{ ' and ' }`

robertsreberski replied:
That's a good preference, thanks!

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/index.tsx:
```diff
@@ -0,0 +1,44 @@
+	const [ viewType, setViewType ] = useState< ConsentViewType >( ConsentViewType.Simple );
+	const handleSetCustomizedView = () => {
+		setViewType( ConsentViewType.Customized );
+	};
```
I am not sure this is a huge deal, but I think it would be nice to have an ability to switch back to the Simple view instead of locking it in the Customized view

gmovr replied:
I was thinking about this, but I think once we have the user choose cutomize (to set custom cookie preferences), they still have the option to choose all cookies, so I don't think there's a need to "go back" to the collapsed cookie banner. We could add it though, as a follow-up, it's an easy change. :)

robertsreberski replied:
We leave it as-is in this PR, and we can always come back to it in the future. From the perspective of UX, I don't think we need to create a "go back" button, as everything is still achievable from the "customized" view.

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/styles.scss:
```diff
@@ -0,0 +1,256 @@
+$cookie-banner-bg-color: #FFFFFF;
+$cookie-banner-text-color: #000000;
+$cookie-banner-text-muted-color: #444444;
+$cookie-banner-link-color: #117ac9;
```
Extreme Nitpick Alert: probably should stick to either all capital letters or all lowercase when writing hex codes 😄

robertsreberski replied:
I'll take care of this when adjusting the code to scss ruleset requirements in the repo, thanks!

robertsreberski replied:
It's actually enforced by stylelint, all changed to lowercase

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/styles.scss:
```diff
@@ -0,0 +1,256 @@
+	width:100%;
+	margin: 25px;
+	bottom: 25px;
+	margin: 0;
```
Seems as though the margin is being set twice, once at 25px 2 lines up, and here at 0

gmovr replied:
I think that's my bad, I'll fix it on the WPcom differential as well. 👍

robertsreberski replied:
I'll clean this up after merging the newest changes from [WPcom diff](D85697-code).

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/styles.scss:
```diff
@@ -0,0 +1,256 @@
+	margin: 25px;
+	bottom: 25px;
+	margin: 0;
+	max-height: 90vh;
```
Quick question, why is the max height set at 90vh? Apologies if the answer is obvious

gmovr replied:
Initially I set it to `100vh`, but then I realized that it's more user friendly if it appears as a modal and doesn't cover the whole screen vertically. So the user can still see where they are, and not hit the back button.

CodeyGuyDylan replied:
Got it, thanks!

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/styles.scss:
```diff
@@ -0,0 +1,256 @@
+		width: 75%;
+		letter-spacing: 0.24px;
+		color: $cookie-banner-text-muted-color;
+		font-weight: 400;
```
I don't think this is needed as it is the default font-weight for <p> tags

gmovr replied:
Probably not needed. This is also some remnants from the WPcom CSS. I tried to set all the CSS as explicit as possible, since this will run across multiple properties (Calypso, WPcom, Jetpack.com, etc.). To make sure that the styles are specific enough to not be affected by any CSS set in the property itself. We had some issues in Calypso, and also across WPcom blogs, where the parent page styles would affect the cookie banner fonts/alignments/colors.

robertsreberski replied:
I'll take care of this during the style cleanup.

CodeyGuyDylan replied:
Makes sense to have it if there is concern of being overridden by priority elsewhere. If that is a legitimate concern I see no issue with it 😄

CodeyGuyDylan reviewed (changes requested):
Overall this looks great! I love the storybook integration! I left some thoughts below, nothing major 😄 

I am seeing on narrower screens that the text on either side are touching the sides of the screen, I am not sure if this has something to do with how storybook is rendering the component? Thought I'd point it out
![image](https://user-images.githubusercontent.com/65001528/203433134-beaddf99-6326-4a9d-a61a-7ada2614ceab.png)

gmovr commented:
> Overall this looks great! I love the storybook integration! I left some thoughts below, nothing major 😄
> 
> I am seeing on narrower screens that the text on either side are touching the sides of the screen, I am not sure if this has something to do with how storybook is rendering the component? Thought I'd point it out ![image](https://user-images.githubusercontent.com/65001528/203433134-beaddf99-6326-4a9d-a61a-7ada2614ceab.png)

This is something I set in the CSS. If the viewport is narrow, the cookie banner will go to full width. This is done to prevent having vertical scroll as much as possible. I also thought it looked better, to not squeeze the cookie banner together from the sides.

I also left some comments on the inline comments regarding the CSS. I definitely think we can iterate here and change styles down the line, also in terms of the wording (we can probably shorten the content further).

gmovr commented:
I checked out the branch locally, and it works well. We should investigate the failing code style tests (I think the E2E will pass after rebase). 👍  There are some minor style changes we can copy over from D85697-code. It's mostly related to typography and some minor changes for mobile alignment.

robertsreberski committed:
Merge remote-tracking branch 'origin/trunk' into add/privacy-components-lib

robertsreberski commented:
Thanks to you both @CodeyGuyDylan and @gmovr for the review on this component. I'm proceeding with merging the recent styles changes from D85697-code and then cleaning up the file - according to your comments and the ruleset running on the repo.

robertsreberski committed:
Merge the most recent styles from wpcom diff, and resolve stylelint errors

robertsreberski committed:
Improve description of library

robertsreberski committed:
Improve importing types

robertsreberski committed:
Replace legal text with lorem ipsum

robertsreberski committed:
Merge remote-tracking branch 'origin/trunk' into add/privacy-components-lib

robertsreberski commented:
> This is something I set in the CSS. If the viewport is narrow, the cookie banner will go to full width. This is done to prevent having vertical scroll as much as possible. I also thought it looked better, to not squeeze the cookie banner together from the sides.

@gmovr all styles are correct, the issue is we have set `box-sizing` on the element `.cookie-banner` itself was set to `content-box` - so it was "ignoring" padding. Setting it to `border-box` fixes all issues.

robertsreberski committed:
Fix width of cookie banner on mobile devices

robertsreberski commented:
I've addressed all comments, merged the recent styles, and cleaned up the code a bit.

I think we can still improve the mobile styles of this cookie banner and clean styles even more. The question is if we want to do it as part of this PR or later after the initial version will be published.

jboland88 commented on packages/privacy-toolset/src/cookie-banner/index.tsx:
```diff
@@ -0,0 +1,44 @@
+		setViewType( ConsentViewType.Customized );
+	};
+	const handleAcceptAll = useCallback( () => {
+		onAccept( { essential: true, analytics: true, advertising: true } );
```
Would it be helpful to define this default for AcceptAll as a `const`? Not a blocker, but noticed this was done elsewhere and it seemed like a nice method as it allowed for the variable to be typed.

jboland88 commented on packages/privacy-toolset/README.md:
```diff
@@ -0,0 +1,37 @@
+
+const Component = () => (
+	<>
+		<CookieBanner content={ ...args } onAccept={ fn } />
```
Since the required props are nested, providing an example of the structure here would be nice for those implementing the package. I think the storybook component could serve this purpose, could potentially reference that here.

robertsreberski replied:
Especially as the CookieBanner component is fully typed, I'm not sure if that's necessarily needed - I used this example mainly to explain how to import components from this library in general. However, I agree that some disclaimers will clarify where to seek more info on param definitions.

jboland88 reviewed (commented):
This tests as described! Looks really nice, easy to follow.

The only thing I'd push for here is adding a few unit tests to prove that the controls do predictable things in case there is an edit in the future that breaks the predicted behavior.

Noted a couple other non-blockers.

I don't have much experience with setting up a package.json for a new package like this, what you have looks reasonable, but that's the one bit that I don't have useful feedback on.

robertsreberski committed:
Implement jest in library

robertsreberski committed:
Add better support for testing

robertsreberski committed:
Add CookieBanner tests and snapshot

robertsreberski committed:
Move allBucketsTrue to reusable consts

robertsreberski committed:
Inform about the detailed content object structure definition

robertsreberski committed:
Improve typing of GranularConsent

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/granular-consent.tsx:
```diff
@@ -23,10 +19,15 @@ export const GranularConsent = ( {
+				onChange={ handleChange }
+				checked={ checked }
+				disabled={ disabled }
+				data-testid={ `${ name }-bucket-toggle` }
```
I have no strong opinions of my own (yet) but have heard very strong opinions that components should not know they are being tested, i.e. nothing about the component exists solely to help a test out. I also found [this article](https://betterprogramming.pub/why-you-should-avoid-testing-react-components-with-test-ids-ee50d20d37d2) explaining why you don't have to use `data-testid` in most situations

Like I said, I have no strong opinions yet as I am pretty new to testing, just wanted to note this in case is changed how you want the tests to work at all 😄

robertsreberski replied:
In general, I agree that we should restrain from using `data-testid` as it's a "shortcut". However, in this case, the input is pretty custom (i.e. hard to identify because of its HTML structure) so this kind of testid helped me to focus on the core of the test, rather than writing the long selector (I would need to find an input that is a sibling of a text).

It's not my go-to approach but I believe this is one of the exceptions we can actually depend on `data-testid`.

jboland88 commented on packages/privacy-toolset/src/cookie-banner/test/index.tsx:
```diff
@@ -0,0 +1,91 @@
+		const { container } = render(
+			<CookieBanner content={ genericContent } onAccept={ jest.fn } />
+		);
+		expect( container ).toMatchSnapshot();
```
What do you think about checking for the presence of key controls here? Maybe seeing if the Accept and Customize buttons exist? The snapshot will tell us if any of the output changes, but we can always regenerate the snapshot to make the test pass.

robertsreberski replied:
> but we can always regenerate the snapshot to make the test pass.

same as you can always change a test to make it pass. I think of them as good indicators you should pay attention to whether you're impacting something visually - to ensure the change is actually expected.

> Maybe seeing if the Accept and Customize buttons exist?

hmm, I'm not sure about it - it will not hurt to add these checks, but they are not necessary IMO

I was always thinking that in moments you need to check multiple visual elements (like the existence of paragraphs and buttons) - matching snapshot is a way to go. but I'm open to the discussion and improving these test cases so to make this component more bulletproof.

jboland88 replied:
> I was always thinking that in moments you need to check multiple visual elements (like the existence of paragraphs and buttons) - matching snapshot is a way to go

Yeah, I think snapshot tests are useful here for making sure future versions of the output don't change. Seems worth including here.

I've always thought of snapshot tests as fragile since any changes have to be manually checked when there are differences. It's possible that a developer could make a known change, expect the snapshot to be different and quickly regenerate the snapshot without realizing that their known change created another unknown effect. 

As you mentioned, it's also not super useful to check that a bunch of different sub-components are rendered/ exist. Especially when the main goal of a test is to prove that a component renders successfully. Perhaps we don't need to check for the Accept/ Customize buttons in this test since you are effectively testing that those buttons exist and are actionable in later tests. If this were the _only_ test in the file, I'd lean towards testing more than a snapshot, but I think just a snapshot is OK given that the later tests prove the function of the critical controls.

robertsreberski replied:
I understand your point, thanks for sharing it. In that way, I'll actually add a test to ensure the critical controls are in order, and the rest of the elements I'll leave up to the snapshot test.

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/test/index.tsx:
```diff
@@ -0,0 +1,91 @@
+ */
+import { render, fireEvent, getByText, getByTestId } from '@testing-library/react';
+import '@testing-library/jest-dom';
+import { CookieBanner, CookieBannerProps } from '..';
```
Looks like we can `import type` here for the `CookieBannerProps` item 😄

robertsreberski replied:
Well, it's a test file, but I take a point we should have it a default. I'll try to figure out how to make auto-imports learn that as well 😅

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/test/index.tsx:
```diff
@@ -0,0 +1,91 @@
+			advertising: true,
+		} );
+	} );
+} );
```
I like all the test cases! Nice!

CodeyGuyDylan commented on packages/privacy-toolset/src/cookie-banner/styles.scss:
```diff
@@ -0,0 +1,265 @@
+			position: absolute;
+			left: 4px;
+			top: 4px;
+			border-radius: $handle-size / 2;
```
Getting a SASS warning here, could be worth updating
![image](https://user-images.githubusercontent.com/65001528/204636993-eba4fcce-9612-4978-af09-508db015815a.png)

robertsreberski replied:
That's true, let's be fully up-to-date in the new code!

CodeyGuyDylan reviewed (approved):
I made a couple of comments/suggestions, but I don't think any of them are blockers. Everything looks and works well, LGTM 😄

robertsreberski committed:
Update tests and README about them

robertsreberski committed:
Update sass divide function

robertsreberski committed:
Revert dedupe on yarn.lock

robertsreberski commented:
Thanks again @jboland88 and @CodeyGuyDylan for your reviews! I've addressed your feedback and prepared the PR for shipping. I plan to do it tomorrow morning incl. publishing it to NPM.

robertsreberski committed:
Remove redundant packages

robertsreberski committed:
Add @automattic/calypso-build
